---
title: "Bistro2: analysis of amplicon bisulfite sequencing"
author: "Mark Ziemann"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
theme: cosmo
---

## Intro

```{r,libs}
suppressPackageStartupMessages({
  library("reshape2")
  library("kableExtra")
  library("dplyr")
  library("RColorBrewer")
  library("GenomicRanges")
  library("limma")
  library("methylKit")
  library("gplots")
  library("seqinr")
})

```

## Notes

Pool1 and pool1 data were merged before analysis.

There is some data named pool14b which was changed to pool14.

My notes from our meeting yesterday:

* Percent CpG sites with usable data (done)

* Just provide the beta values (done)

* remove pool8, pool17 and rerun stats

* Adjust by bisulfite conversion

## Off target reads

```{r, offtarget, fig.height=8,fig.width=8}

DATADIR="fastq"

oft_name = paste(DATADIR,"/3col_offtarget.tsv",sep="")

oft <- read.table(oft_name)

x <- t( as.matrix(acast(oft, V2~V1, value.var="V3")) )

x <- x[,which(colSums(x)!=0)]

x %>% kbl() %>% kable_styling()

xx <- t(x/rowSums(x) * 100)

t(xx) %>% kbl() %>% kable_styling()

par(mar=c(5, 10, 5, 2))
barplot(rowSums(x),horiz=TRUE,las=1,cex.names=0.7,main="total reads") ; grid()

barplot(x[,1],horiz=TRUE,las=1,cex.names=0.7,main="assigned reads (on target)") ; grid()

barplot(x[,2],horiz=TRUE,las=1,cex.names=0.7,main="unassigned reads (ambiguity)") ; grid()

barplot(x[,3],horiz=TRUE,las=1,cex.names=0.7,main="unassigned reads (off target)") ; grid()

barplot(x[,4],horiz=TRUE,las=1,cex.names=0.7,main="unmapped reads ") ; grid()

par(mar=c(5, 10, 5, 2))
barplot(xx,horiz=TRUE,las=1,cex.names=0.6,xlab="proportion of reads (%)", 
  legend.text=TRUE,
  args.legend = list(x = "topright" , bty = "n", inset=c(0, -0.15), cex=0.8))
  grid()

barplot(xx[1,],horiz=TRUE,las=1,cex.names=0.7,main="proportion of reads on target (%)") ; grid()

```


## On target reads

```{r, ontarget, fig.height=8,fig.width=8}

ont_name = paste(DATADIR,"/3col_ontarget.tsv",sep="")

ont <- read.table(ont_name)

y <-t( as.matrix(acast(ont, V2~V1, value.var="V3",fun.aggregate=sum)) )

y %>% kbl() %>% kable_styling()

yy <- t(y/rowSums(y) * 100)

par(mar=c(5, 10, 5, 2))
barplot(yy,horiz=TRUE,las=1,cex.names=0.6,xlab="proportion of reads (%)",
  legend.text=TRUE,col=brewer.pal(n = 7, name = "Set1"),
  args.legend = list(x = "left" , bty = "n", inset=c(-0.3, -0.15), cex=0.8))
grid()

yy <- yy[,which( ! colnames(yy) %in% c("Neg","101-1023","185-1135","7185","NTC","Pool1")   )]

colfunc <- colorRampPalette(c("white", "blue"))

heatmap.2(yy,trace="none",scale="none",col=colfunc(25))


```

## Look at on-target reads

Here we have a summary of the reads that were on target for each amplicon for each individual.

```{r, ontarget2}

myfiles <- list.files("fastq/",pattern="bedgraph.intersect",full.names=TRUE)

ont <- lapply(myfiles, function(x)  { y <- read.table(x,header=TRUE) ; y  }  )

names(ont) <- gsub(".bam.bedgraph.intersect","",gsub("fastq//","",myfiles))

ont

```

## Look at off-target reads

Here we have the details of each of the off-target genomic regions that were identified in this assay.
I've also got the location of the closest gene.

```{r, offtarget2}

myfiles <- list.files("fastq/",pattern=".bedgraph.offtarget.genes$",full.names=TRUE)

oft <- lapply(myfiles, function(x)  { y <- read.table(x,header=TRUE) ; y  }  )

names(oft) <- gsub(".bam.bedgraph.offtarget.genes","",gsub("fastq//","",myfiles))

oft

```

Now we can look at the sequence of these off-target regions. 
Lower case indicates that the sequence was repeat-masked.

```{r, offtarget3}

myfiles <- list.files("fastq/",pattern=".bedgraph.offtarget.fa$",full.names=TRUE)

oft2 <- lapply(myfiles, function(x)  {
  y <- readLines(x)
  n=length(y)
  if(n>0){
    seq <- y[(1:(n/2))*2]
    names(seq) <- y[((1:(n/2))*2)-1]
    seq  
  }
})

names(oft2) <- gsub(".bam.bedgraph.offtarget.fa","",gsub("fastq//","",myfiles))

oft2

```

## Look at the origin of the off-target reads

If we look at the similarity of the genomic locations with the amplicons we should be able to track down which primer was 
responsible for which off-target region.
The primers are 16-26 bp in length so I will use the 20bp at the termini to do this classification.

In these heatmaps, the seq names along the bottom are the target amplicons and the ones along the right are the off-target regions.

The deeper the red, the smaller the sequence edit distance.

```{r, offtarget4,fig.width=8,fig.height=8}

# amplicon termini

x <- readLines("amplicons_hg38.bed.starts.fa")
n=length(x)
amp_st <- x[(1:(n/2))*2]
names(amp_st) <- x[((1:(n/2))*2)-1]

x <- readLines("amplicons_hg38.bed.ends.fa")
n=length(x)
amp_en <- x[(1:(n/2))*2]
names(amp_en) <- x[((1:(n/2))*2)-1]

amp <-c(amp_st,amp_en)

# seq termini
mystarts <- list.files("fastq/",pattern=".bedgraph.offtarget.starts.fa$",full.names=TRUE)
myends <- list.files("fastq/",pattern=".bedgraph.offtarget.ends.fa$",full.names=TRUE)

idxs <- grep("ool",mystarts)

lapply(idxs, function(z) {
  mystart <- mystarts[z]
  x <- readLines(mystart)
  n=length(x)
  st <- x[(1:(n/2))*2]
  names(st) <- x[((1:(n/2))*2)-1]
  myend <- myends[z]
  x <- readLines(myend)
  n=length(x)
  en <- x[(1:(n/2))*2]
  names(en) <- x[((1:(n/2))*2)-1]
  seq <-c(st,en)
  adist(seq,amp)
  colfunc <- colorRampPalette(c("red", "white"))
  heatmap.2(adist(seq,amp),col=colfunc(50),trace="none",scale="none",margin=c(15,20),main=names(oft2)[z])
})

```

## Look at methylation levels

Looking at the average CpG methylation levels across the whole genome.
It shows that total methylation levels are very high across the whole sample set.
Pool MM is the unmethylated control with ~20% methylation.
Pool M is the methylated control with ~91% methylation.

```{r,cpg_methlevels,fig.height=8,fig.width=8}

myfilelist <- list.files("fastq",pattern="vcf_meth_average.tsv", full.names = TRUE)
samplelist <- gsub(".bam.vcf_meth_average.tsv","",myfilelist)

myread2 <- function(myfile){
  read.table(myfile,header=TRUE)
}

z <- lapply(myfilelist, myread2)
zz <- lapply(z,function(x) {tail(x,1)[4] } )
zz <- unlist(zz)
zz <- gsub("%","",zz)
zz <- as.numeric(zz)

nz <- lapply(z,function(x) {tail(x,1)[1] } )
nz <- unname(gsub(".bam","",gsub("fastq/","",unlist(nz))))
names(zz) <- nz

par(mar=c(5, 10, 5, 2))
barplot(zz,horiz=TRUE,las=1,cex.names=0.6,xlab="percent CpG methylation")
grid()

```

Now I will look at average CpH methylation levels across the whole genome

```{r,cph_methlevels,fig.height=8,fig.width=8}

z <- lapply(myfilelist, myread2)
zz <- lapply(z,function(x) {tail(x,1)[10] } )
zz <- unlist(zz)
zz <- gsub("%","",zz)
zz <- as.numeric(zz)

nz <- lapply(z,function(x) {tail(x,1)[1] } )
nz <- unname(gsub(".bam","",gsub("fastq/","",unlist(nz))))
names(zz) <- nz

par(mar=c(5, 10, 5, 2))
barplot(zz,horiz=TRUE,las=1,cex.names=0.6,xlab="percent CpH methylation")
grid()

```


## Read in data

myobjf looks strange because if the coverage is high, it gets capped to approx 2623.


```{r,readin_qc}

file.list <- list.files(DATADIR,pattern="methylkit.tsv",full.names=TRUE)
lines <- lapply(file.list,readLines)
file.list <- file.list[which(lapply(lines,length) > 0)]

samplesheet <- read.table("sample_info/samplesheet_2020-10-15.tsv",sep="\t",header=TRUE)

samplesheet$poolID <- paste("pool",samplesheet$Internal_ID,sep="")

# filter files for those in the sample sheet
file.list <- sapply(strsplit(file.list,"\\."),"[[",1) 

file.list <- sapply(strsplit(file.list,"/"),"[[",2) 

file.list <- file.list[grep("ool",file.list)]
file.list

myobj <- methRead( location=as.list(paste("fastq/",file.list,".methylkit.tsv",sep="")),
  sample.id=as.list(file.list),
  header=FALSE,
  assembly="hg38",
  treatment=sample(1:length(file.list)),
  context="CpG",
  mincov = 0 )

myobjf <- filterByCoverage(myobj,lo.count=3,lo.perc=NULL,
  hi.count=NULL,hi.perc=NULL)

meth <- unite(myobjf, destrand=FALSE)

percMethylation(meth) %>% kbl() %>% kable_styling()


```

## Export data for downstream analysis

Here I'm exporting the number of Cs and Ts at each position.
Then I'm also providing the beta values.

```{r,export_meth_data}

mytable <- getData(meth)

write.table(x=mytable,file="methylation_data.tsv",sep="\t",quote=FALSE)

mytable2 <- percMethylation(meth)

mytable3 <- cbind(mytable[,1:3],mytable2)

write.table(x=mytable3,file="methylation_betavals.tsv",sep="\t",quote=FALSE)

```

## Proportion of CpG sites with usable data

* Percent CpG sites with usable data

```{r,usable_data_cg}

fa <- read.fasta("amplicons_hg38.bed.fa")

# count CpG sites
cnt_cpg <- function(fa1) {
  length(which(
    sapply( 1:(length(fa1)-1), function(i) {
      motif <- fa1[i:(i+1)]
      sum(as.numeric(motif == c("c","g")) )==2  
    })
  ))
}

num_cpg <- lapply(fa, cnt_cpg)

start_pos <- sapply(strsplit(names(num_cpg),":"),"[[",2)
start_pos <- sapply(strsplit(start_pos,"-"),"[[",1)

end_pos <- sapply(strsplit(names(num_cpg),":"),"[[",2)
end_pos <- sapply(strsplit(end_pos,"-"),"[[",2)

chrs <- sapply(strsplit(names(num_cpg),":"),"[[",1)

res <- lapply(1:length(num_cpg),function(i) {
  mychr <- chrs[i]
  mysubset <- mytable3[ which(mytable3$chr==mychr),]
  mystart <- start_pos[i]
  mysubset <- mysubset[which(mysubset$start > mystart),]
  myend <- end_pos[i]
  mysubset <- mysubset[which(mysubset$end < myend),]
  nrow(mysubset)
})

mydf <- as.data.frame(unlist(num_cpg))

mydf$covered_cpg <- unlist(res)

colnames(mydf)[1] <- "num_CpGs"

mydf %>% kbl() %>% kable_styling()


```

## Basic plots

```{r, plots1, fig.height=8,fig.width=8}

getCorrelation(meth,plot=TRUE)
clusterSamples(meth, dist="correlation", method="ward", plot=TRUE)
PCASamples(meth, screeplot=TRUE)
pc <- PCASamples(meth,obj.return=TRUE)
plot(pc$x[,1],pc$x[,2],bty="n",type="n")
grid()
text(pc$x[,1],pc$x[,2],labels=rownames(pc$x),cex=0.7)

```

Now to remove control samples and others not included on the samplesheet.


```{r,readin_qc2}

file.list <- file.list[which(file.list %in% samplesheet$poolID)]

# filter samplesheet for those with files
samplesheet <- samplesheet[which( samplesheet$poolID  %in% file.list ),]

myobj <- methRead( location=as.list(paste("fastq/",file.list,".methylkit.tsv",sep="")),
  sample.id=as.list(samplesheet$poolID),
  header=FALSE,
  assembly="hg38",
  treatment=samplesheet$GWG,
  context="CpG",
  mincov = 5 )

myobjf <- filterByCoverage(myobj,lo.count=5,lo.perc=NULL,
  hi.count=NULL,hi.perc=NULL)

meth <- unite(myobjf, destrand=FALSE)

```

## Read amplicon features

```{r,amplicons}

bed <- read.table("amplicons_hg38.bed")

amp <- GRanges(seqnames=bed$V1,ranges=IRanges(bed$V2,bed$V3))

amp$gene_names <- bed$V4

```

## Statistical analysis

```{r,stats}

dm <- calculateDiffMeth(meth)

dm

dm$pvalue<0.05

min(dm$pvalue)

plot(dm$meth.diff,-log10(dm$pvalue) )

hist(dm$meth.diff)

#dm.hyper=getMethylDiff(dm,difference=1,qvalue=0.01,type="hyper")

#dm.hypo=getMethylDiff(dm,difference=1,qvalue=0.01,type="hypo")

```

## Lets run a simple correlation test

Here we run a spearman correlation test.

```{r,correl1}

z <- samplesheet$GWG

genes <- amp

genes %>% kbl() %>% kable_styling()

par(mfrow=c(1,2))
x <- lapply( 1:nrow(meth), function(i) {
  x <- as.vector(percMethylation(meth)[i,])
  cor <- cor.test(x,z,method="pearson")
  p <- cor$p.value
  r <- cor$estimate
  CHR=as.vector(seqnames(GRanges(meth)))[i]
  POS=as.character(start(ranges(GRanges(meth)))[i])
  CHRPOS=paste("chr",CHR,":",POS,sep="")
  HEADER = paste("Site",i,CHRPOS,"p=",round(p,3)," r=",round(r,3))

  plot(z,x,xlab="GWG",ylab="Beta",main="Pearson")
  mtext(HEADER)
  mylm <- lm( x ~ z )
  abline(mylm,lwd=3,lty=2,col="red")

  cor <- cor.test(x,z,method="spearman")
  p <- cor$p.value
  r <- cor$estimate
  CHR=as.vector(seqnames(GRanges(meth)))[i]
  POS=as.character(start(ranges(GRanges(meth)))[i])
  CHRPOS=paste("chr",CHR,":",POS,sep="")
  HEADER = paste("Site",i,CHRPOS,"p=",round(p,3)," r=",round(r,3))

  plot(rank(z),rank(x),xlab="rank(GWG)",ylab="rank(Beta)",main="Spearman")
  mtext(HEADER)
  mylm <- lm( rank(x) ~ rank(z) )
  abline(mylm,lwd=3,lty=2,col="red")

  c("p"=p,"r"=r)
})

```

Now let's present the results.

```{r,correl2}

xx <- as.data.frame(t(sapply(x,function(x){ x[c(2,1)] } )))
xx$FDR <- p.adjust(xx[,2],method="fdr")
xx <- data.frame(meth$chr,meth$start,xx)
colnames(xx) <- c("seqname","pos","rho","p","FDR")

gr <- GRanges(seqnames = xx$seqname, ranges = IRanges(xx$pos) )
gr$r <- xx$r
gr$p <- xx$p
gr$FDR <- xx$FDR

ol <- findOverlaps(gr,genes)

# off target gene methylation
ot <- gr[-queryHits(ol)]
ot %>% kbl() %>% kable_styling()

# on target methylation
gr <- gr[queryHits(ol)]

gr$gene_names <- genes$gene_names[subjectHits(ol) ]

gr %>% kbl() %>% kable_styling()

```

## Remove outliers

Two samples (pool8, pool17) showed unusual profiles that are consistent with incomplete bisulfite conversion.

```{r,remove_pool8_pool17}

file.list <- file.list[which(! file.list %in% c("pool8","pool17"))]

samplesheet <- samplesheet[which(! samplesheet$poolID %in% c("pool8","pool17")),]


myobj <- methRead( location=as.list(paste("fastq/",file.list,".methylkit.tsv",sep="")),
  sample.id=as.list(samplesheet$poolID),
  header=FALSE,
  assembly="hg38",
  treatment=samplesheet$GWG,
  context="CpG",
  mincov = 5 )

myobjf <- filterByCoverage(myobj,lo.count=5,lo.perc=NULL,
  hi.count=NULL,hi.perc=NULL)

meth <- unite(myobjf, destrand=FALSE)

```

Now run a spearman correlation test.

```{r,correl3}

z <- samplesheet$GWG

genes <- amp

genes %>% kbl() %>% kable_styling()

par(mfrow=c(1,2))
x <- lapply( 1:nrow(meth), function(i) {
  x <- as.vector(percMethylation(meth)[i,])
  cor <- cor.test(x,z,method="pearson")
  p <- cor$p.value
  r <- cor$estimate
  CHR=as.vector(seqnames(GRanges(meth)))[i]
  POS=as.character(start(ranges(GRanges(meth)))[i])
  CHRPOS=paste("chr",CHR,":",POS,sep="")
  HEADER = paste("Site",i,CHRPOS,"p=",round(p,3)," r=",round(r,3))

  plot(z,x,xlab="GWG",ylab="Beta",main="Pearson")
  mtext(HEADER)
  mylm <- lm( x ~ z )
  abline(mylm,lwd=3,lty=2,col="red")

  cor <- cor.test(x,z,method="spearman")
  p <- cor$p.value
  r <- cor$estimate
  CHR=as.vector(seqnames(GRanges(meth)))[i]
  POS=as.character(start(ranges(GRanges(meth)))[i])
  CHRPOS=paste("chr",CHR,":",POS,sep="")
  HEADER = paste("Site",i,CHRPOS,"p=",round(p,3)," r=",round(r,3))

  plot(rank(z),rank(x),xlab="rank(GWG)",ylab="rank(Beta)",main="Spearman")
  mtext(HEADER)
  mylm <- lm( rank(x) ~ rank(z) )
  abline(mylm,lwd=3,lty=2,col="red")

  c("p"=p,"r"=r)
})

```

```{r,correl4}

xx <- as.data.frame(t(sapply(x,function(x){ x[c(2,1)] } )))
xx$FDR <- p.adjust(xx[,2],method="fdr")
xx <- data.frame(meth$chr,meth$start,xx)
colnames(xx) <- c("seqname","pos","rho","p","FDR")

gr <- GRanges(seqnames = xx$seqname, ranges = IRanges(xx$pos) )
gr$r <- xx$r
gr$p <- xx$p
gr$FDR <- xx$FDR

ol <- findOverlaps(gr,genes)

# off target gene methylation
ot <- gr[-queryHits(ol)]
ot %>% kbl() %>% kable_styling()

# on target methylation
gr <- gr[queryHits(ol)]

gr$gene_names <- genes$gene_names[subjectHits(ol) ]

gr %>% kbl() %>% kable_styling()

```

## Adjust by bisulfite conversion

## Session info

```{r,sessioninfo}

sessionInfo()

```



